File: core/distributed_extraction/distributed_extraction_orchestrator.py
from __future__ import annotations

from typing import Any, Dict, List, Optional

from core.distributed_extraction.distributed_adaptive_runtime_engine import (
    synchronize_adaptive_runtime,
)
from core.distributed_extraction.distributed_cluster_engine import (
    build_cluster_state,
)
from core.distributed_extraction.distributed_identity_engine import (
    route_browser_identity,
)
from core.distributed_extraction.distributed_monitoring_engine import (
    monitor_extraction_cluster,
)
from core.distributed_extraction.distributed_recovery_engine import (
    recover_distributed_runtime,
)
from core.distributed_extraction.distributed_runtime_graph_engine import (
    build_distributed_runtime_graph,
)
from core.distributed_extraction.distributed_session_engine import (
    route_authenticated_sessions,
)
from core.distributed_extraction.distributed_stream_engine import (
    federate_stream_runtimes,
)
from core.distributed_extraction.extraction_queue_engine import (
    dequeue_extraction,
    enqueue_extraction,
)
from core.distributed_extraction.extraction_scheduler_engine import (
    schedule_extraction_runtime,
)
from core.distributed_extraction.extraction_worker_engine import (
    create_extraction_worker,
)
from core.distributed_extraction.runtime_federation_engine import (
    federate_extraction_runtimes,
)
from core.distributed_extraction.distributed_load_balancer import (
    balance_extraction_workloads,
)


def run_distributed_extraction(
    tasks: List[Dict[str, Any]],
    workers: Optional[List[Dict[str, Any]]] = None,
    checkpoint: Optional[Dict[str, Any]] = None,
    tick: int = 0,
    runtime_graphs: Optional[List[Dict[str, Any]]] = None,
) -> Dict[str, Any]:
    checkpoint = dict(checkpoint or {})
    queue = list(checkpoint.get("queue", []))
    worker_list = list(workers or checkpoint.get("workers", []))

    if not worker_list:
        worker_list = [
            create_extraction_worker(
                worker_id="worker_0",
                identity={"profile_id": "default", "fingerprint_hash": "fp0"},
                adaptive_runtime={"memory": {"healed_selectors": {}}},
                stream_runtime={"events": []},
            )
        ]

    for task in tasks:
        queued = enqueue_extraction(queue, task)
        queue = queued["queue"]

    schedule = schedule_extraction_runtime(tasks, tick=tick)
    assignments = balance_extraction_workloads(worker_list, tasks)

    session_routes = route_authenticated_sessions(worker_list)
    identity_routes = route_browser_identity(worker_list)

    adaptive_sync = synchronize_adaptive_runtime(
        [worker.get("adaptive_runtime", {}) for worker in worker_list]
    )

    stream_federation = federate_stream_runtimes(
        [
            {
                "worker_id": worker.get("worker_id"),
                "events": worker.get("stream_runtime", {}).get("events", []),
            }
            for worker in worker_list
        ]
    )

    federation = federate_extraction_runtimes(runtime_graphs or [])
    distributed_graph = build_distributed_runtime_graph(
        worker_list,
        federation.get("topology", {}),
    )

    monitoring = monitor_extraction_cluster(worker_list, queue)
    cluster = build_cluster_state(worker_list, queue)

    next_checkpoint = {
        "queue": queue,
        "workers": worker_list,
        "runtime_graph": distributed_graph,
        "identities": identity_routes.get("routes", []),
        "adaptive_memory": adaptive_sync,
        "stream_runtime": stream_federation,
        "tick": tick + 1,
        "assignments": assignments.get("assignments", []),
        "bounded": True,
    }

    return {
        "workers": worker_list,
        "queue": queue,
        "schedule": schedule,
        "assignments": assignments,
        "session_routes": session_routes,
        "identity_routes": identity_routes,
        "adaptive_sync": adaptive_sync,
        "stream_federation": stream_federation,
        "topology": federation,
        "distributed_graph": distributed_graph,
        "monitoring": monitoring,
        "cluster": cluster,
        "checkpoint": next_checkpoint,
        "bounded": True,
    }

File: src/distributed/distributedExtractionOrchestrator.ts
import { synchronizeAdaptiveRuntime } from "./distributedAdaptiveRuntimeEngine.js";
import { buildClusterState } from "./distributedClusterEngine.js";
import { routeBrowserIdentity } from "./distributedIdentityEngine.js";
import { monitorExtractionCluster } from "./distributedMonitoringEngine.js";
import { buildDistributedRuntimeGraph } from "./distributedRuntimeGraphEngine.js";
import { routeAuthenticatedSessions } from "./distributedSessionEngine.js";
import { federateStreamRuntimes } from "./distributedStreamEngine.js";
import { balanceExtractionWorkloads } from "./distributedLoadBalancer.js";
import { dequeueExtraction, enqueueExtraction } from "./extractionQueueEngine.js";
import { scheduleExtractionRuntime } from "./extractionSchedulerEngine.js";
import { createExtractionWorker } from "./extractionWorkerEngine.js";
import { federateExtractionRuntimes } from "./runtimeFederationEngine.js";

export function runDistributedExtraction(
  tasks: Record<string, unknown>[],
  workers?: Record<string, unknown>[],
  checkpoint: Record<string, unknown> = {},
  tick = 0,
  runtimeGraphs: Record<string, unknown>[] = [],
): Record<string, unknown> {
  let queue = [...((checkpoint.queue as Record<string, unknown>[]) ?? [])];
  let workerList = [...(workers ?? (checkpoint.workers as Record<string, unknown>[]) ?? [])];

  if (!workerList.length) {
    workerList = [
      createExtractionWorker("worker_0", {}, { profile_id: "default", fingerprint_hash: "fp0" }, {
        memory: { healed_selectors: {} },
      }, { events: [] }),
    ];
  }

  for (const task of tasks) {
    queue = enqueueExtraction(queue, task).queue;
  }

  const schedule = scheduleExtractionRuntime(tasks, tick);
  const assignments = balanceExtractionWorkloads(workerList, tasks);
  const sessionRoutes = routeAuthenticatedSessions(workerList);
  const identityRoutes = routeBrowserIdentity(workerList);
  const adaptiveSync = synchronizeAdaptiveRuntime(
    workerList.map((w) => (w.adaptive_runtime as Record<string, unknown>) ?? {}),
  );
  const streamFederation = federateStreamRuntimes(
    workerList.map((w) => ({
      worker_id: w.worker_id,
      events: ((w.stream_runtime as Record<string, unknown>)?.events as unknown[]) ?? [],
    })),
  );
  const federation = federateExtractionRuntimes(runtimeGraphs);
  const distributedGraph = buildDistributedRuntimeGraph(
    workerList,
    federation.topology as Record<string, unknown>,
  );
  const monitoring = monitorExtractionCluster(workerList, queue);
  const cluster = buildClusterState(workerList, queue);

  const nextCheckpoint = {
    queue,
    workers: workerList,
    runtime_graph: distributedGraph,
    identities: identityRoutes.routes as unknown[],
    adaptive_memory: adaptiveSync,
    stream_runtime: streamFederation,
    tick: tick + 1,
    assignments: assignments.assignments as unknown[],
    bounded: true,
  };

  return {
    workers: workerList,
    queue,
    schedule,
    assignments,
    session_routes: sessionRoutes,
    identity_routes: identityRoutes,
    adaptive_sync: adaptiveSync,
    stream_federation: streamFederation,
    topology: federation,
    distributed_graph: distributedGraph,
    monitoring,
    cluster,
    checkpoint: nextCheckpoint,
    bounded: true,
  };
}

export { dequeueExtraction, enqueueExtraction, createExtractionWorker };
